src/expr.jl

```diff
@@ -28,3 +28,157 @@ function extract_kwargs(expr::Expr)
+    end
+
+    # long-form `:function`, short-form `:(=)`, anonymous `:(->)`
+    if !(ex.head === :function || ex.head === :(=) || ex.head === :(->))
+        return invalid_def("invalid function type `$(repr(ex.head))`")
+    end
+
+    def[:type] = ex.head
+    anon = ex.head == :(->)
+
+    if ex.head === :function && length(ex.args) == 1  # empty function definition
+        def[:name] = ex.args[1]
+        return def
+    elseif length(ex.args) == 2  # Expect signature and body
+        def[:body] = ex.args[2]
+        ex = ex.args[1]  # Focus on the function signature
+    else
+        quan = length(ex.args) > 2 ? "too many" : "too few"
+        return invalid_def("$quan of expression arguments for `$(repr(def[:type]))`")
+    end
+
+    # Where parameters
+    if ex isa Expr && ex.head === :where
+        def[:whereparams] = Any[]
+
+        while ex isa Expr && ex.head === :where
+            append!(def[:whereparams], ex.args[2:end])
+            ex = ex.args[1]
+        end
+    end
+
+    # Return type
+    if !anon && ex isa Expr && ex.head === :(::)
+        def[:rtype] = ex.args[2]
+        ex = ex.args[1]
+    end
+
+    # Arguments and keywords
+    if ex isa Expr && (anon && ex.head === :tuple || !anon && ex.head === :call)
+        i = anon ? 1 : 2
+
+        if length(ex.args) >= i
+            if ex.args[i] isa Expr && ex.args[i].head === :parameters
+                def[:kwargs] = ex.args[i].args
+
+                if length(ex.args) > i
+                    def[:args] = ex.args[(i + 1):end]
+                end
+            else
+                def[:args] = ex.args[i:end]
+            end
+        end
+    elseif anon
+        def[:args] = [ex]
+    else
+        return invalid_def("invalid or missing arguments")
+    end
+
+    # Function name and type parameters
+    if !anon
+        ex = ex.args[1]
+
+        if ex isa Expr && ex.head === :curly
+            def[:params] = ex.args[2:end]
+            ex = ex.args[1]
+        end
+
+        def[:name] = ex
+    end
+
+    return def
+end
+
+function combinedef(def::Dict{Symbol,Any})
+    # Determine the name of the function including parameterization
+    name = if haskey(def, :params)
+        Expr(:curly, def[:name], def[:params]...)
+    elseif haskey(def, :name)
+        def[:name]
+    else
+        nothing
+    end
+
+    # Empty generic function definitions must not contain additional keys
+    if !haskey(def, :body) && any(k -> haskey(def, k), [:args, :kwargs, :rtype, :whereparams])
+        extra = [:args, :kwargs, :rtype, :whereparams]
+        throw(ArgumentError(string(
+            "Function definitions without a body must not contain keys: ",
+            join(string.('`', repr.(setdiff(extra, keys(def))), '`'), ", ", ", or "),
+        )))
+    end
+
+    # Combine args and kwargs
+    args = Any[]
+    !isempty(get(def, :kwargs, [])) && push!(args, Expr(:parameters, def[:kwargs]...))
+    !isempty(get(def, :args, [])) && append!(args, def[:args])
```

Comment:
For all in intents and purposes  the above and

```suggestion
    haskey(def, :args)) && append!(args, def[:args])
```

or even 

```suggestion
    append!(args, get(def, :args, []))
```

are going to be the same.

And I think they are clearer than 
```
!isempty(get(def, :args, [])) && append!(args, def[:args])
```

Reply:
After reading this I have been adding some tests for `(;) -> nothing` which seems to be a corner case I wasn't previously testing for.

Reply:
`haskey` works here.